File: 2023/haskell/day_1.hs
{-# LANGUAGE QuasiQuotes #-}

import Test.HUnit
import Text.RawString.QQ

import Data.Char
import Data.List

import Lib

main = do
  input <- fmap parseInput (readFile "../input/day_1.txt")
  runTestTT tests
  print . part1 $ input
  print . part2 $ input

parseInput = lines . trim

firstAndLast digits = 10 * head digits + last digits

part1 :: [String] -> Int
part1 input = sum (map (firstAndLast . map digitToInt . filter isDigit) input)

digitStrings :: [String]
digitStrings =
  ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

findDigitString line =
  let findPrefix' line [] = Nothing
      findPrefix' line ((i, x):xs)
        | x `isPrefixOf` line = Just i
        | otherwise = findPrefix' line xs
   in findPrefix' line (zip [1 ..] digitStrings)

parseDigits [] = []
parseDigits line@(x:xs)
  | isDigit x = digitToInt x : parseDigits xs
  | Just d <- findDigitString line = d : parseDigits xs
  | otherwise = parseDigits xs

part2 :: [String] -> Int
part2 input = sum (map (firstAndLast . parseDigits) input)

example1 =
  [r|
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
|]

part1_test = TestCase (assertEqual "example1" (part1 (parseInput example1)) 142)

example2 =
  [r|
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
|]

part2_test = TestCase (assertEqual "example2" (part2 (parseInput example2)) 281)

tests = TestList [TestLabel "part1" part1_test, TestLabel "part2" part2_test]

File: 2023/python/day_1.py
def parse_input(input):
    return input.strip().splitlines()


def first_and_last(digits):
    return 10 * digits[0] + digits[-1]


def part1(input):
    return sum(first_and_last([int(c) for c in line if c.isdigit()]) for line in input)


digit_strings = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]


def find_digit(s: str, forward: bool) -> int:
    i = 0 if forward else len(s) - 1
    while i >= 0 and i < len(s):
        if s[i].isdigit():
            return int(s[i])
        for j, d in enumerate(digit_strings):
            if s[i:].startswith(d):
                return j + 1
        i += 1 if forward else -1


def first_and_last_part2(line):
    first = find_digit(line, True)
    last = find_digit(line, False)
    return first * 10 + last


def part2(input):
    return sum(first_and_last_part2(line) for line in input)


def main():
    with open("../input/day_1.txt", "r") as f:
        input = parse_input(f.read())

    print("part1:", part1(input))
    print("part2:", part2(input))


if __name__ == "__main__":
    main()


EXAMPLE_1 = """
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
"""


def test_part1():
    assert part1(parse_input(EXAMPLE_1)) == 142


EXAMPLE_2 = """
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
"""


def test_part2():
    assert first_and_last_part2("eight1two3abthreexy") == 83
    assert first_and_last_part2("6sixoneightnc") == 68
    assert part2(parse_input(EXAMPLE_2)) == 281
